python_modules/dagster/dagster/_cli/workspace/cli_target.py

```diff
@@ -354,16 +300,107 @@ class PythonPointerOpts:
     working_directory: Optional[str] = None
     attribute: Optional[str] = None
 
+    @classmethod
+    def extract_from_cli_options(cls, cli_options: dict[str, Any]) -> Self:
+        # This is expected to always be called from a click entry point, so all options should be
+        # present in the dictionary. We rely on `@record` for type-checking.
+        return cls(
+            python_file=cli_options.pop("python_file"),
+            module_name=cli_options.pop("module_name"),
+            package_name=cli_options.pop("package_name"),
+            working_directory=cli_options.pop("working_directory"),
+            attribute=cli_options.pop("attribute"),
+        )
+
+    def to_workspace_opts(self) -> "WorkspaceOpts":
+        return WorkspaceOpts(
+            python_file=(self.python_file,) if self.python_file else None,
+            module_name=(self.module_name,) if self.module_name else None,
+            package_name=(self.package_name,) if self.package_name else None,
+            working_directory=self.working_directory,
+            attribute=self.attribute,
+        )
+
+
+@record
+class WorkspaceOpts:
+    empty_workspace: bool = False
+    workspace: Optional[Sequence[str]] = None
+
+    # Like PythonPointerParams but multiple files/modules/packages are allowed
+    python_file: Optional[Sequence[str]] = None
+    module_name: Optional[Sequence[str]] = None
+    package_name: Optional[Sequence[str]] = None
+    working_directory: Optional[str] = None
+    attribute: Optional[str] = None
```

Comment:
* How is attribute not a `Sequence`?
* Also seems like there should a tuple record that has a python_file, module_name, package_name combo as long as we are stucturing this.

Reply:
>How is attribute not a Sequence?

This is just a clear representation of what the API already is. If you pass `attribute` it's an error to pass multiple `python_file` etc. I assume it's this way because there is no way to know which attribute pairs with which module spec:

```
$ ... --python-file foo.py --python-file bar.py --attribute x --attribute y
```

^^ which attribute goes with which file?

Reply:
Got it. Can we structure this data structure to disallow invalid states?

Reply:
Possibly in a follow-up but that would introduce extra complexity here. The invalid states are already caught by existing logic that these flow through. The data structures are designed to be a completely "dumb" representation of what is received via the command line.